File: java/patterns/src/main/java/org/prototypepatterns/behavioral/templatemethod/TemplateMethodPattern.java
package org.prototypepatterns.behavioral.templatemethod;

import java.util.List;

public final class TemplateMethodPattern {
    private TemplateMethodPattern() {
    }

    abstract static class DataPipeline {
        final List<String> run() {
            return List.of(read(), transform(), write());
        }

        abstract String read();

        abstract String transform();

        abstract String write();
    }

    static final class CsvPipeline extends DataPipeline {
        @Override
        String read() {
            return "read:csv";
        }

        @Override
        String transform() {
            return "transform:normalize";
        }

        @Override
        String write() {
            return "write:warehouse";
        }
    }

    public static List<String> demo() {
        return new CsvPipeline().run();
    }
}

File: csharp/patterns/src/PrototypePatterns/Behavioral/TemplateMethod/TemplateMethodPattern.cs
namespace PrototypePatterns.Behavioral.TemplateMethod;

public abstract class DataPipeline
{
    public IReadOnlyList<string> Run() => [Read(), Transform(), Write()];

    protected abstract string Read();
    protected abstract string Transform();
    protected abstract string Write();
}

public sealed class CsvPipeline : DataPipeline
{
    protected override string Read() => "read:csv";

    protected override string Transform() => "transform:normalize";

    protected override string Write() => "write:warehouse";
}

public static class TemplateMethodPattern
{
    public static IReadOnlyList<string> Demo() => new CsvPipeline().Run();
}
